encoding/varint.ts

```diff
@@ -34,22 +38,51 @@ const SHIFTN = 7n;
  * from the returned new `offset`.
  */
 export function decode(buf: Uint8Array, offset = 0): [bigint, number] {
-  for (
-    let i = offset,
-      len = Math.min(buf.length, offset + MaxVarIntLen64),
-      shift = 0,
-      decoded = 0n;
-    i < len;
-    i += 1, shift += SHIFT
-  ) {
-    const byte = buf[i];
-    decoded += BigInt((byte & REST) * Math.pow(2, shift));
-    if (!(byte & MSB) && decoded > MaxUInt64) {
-      throw new RangeError("overflow varint");
+  U64_VIEW[0] = 0n;
+  let intermediate = 0;
+  let position = 0;
+  let i = offset;
+
+  if (buf.length === 0) throw new RangeError("Cannot read empty buffer");
+
+  let byte = buf[i];
+  do {
+    byte = buf[i];
+
+    // 1. Take the lower 7 bits of the byte.
+    // 2. Shift the bits into the correct position.
+    // 3. Bitwise OR it with the intermediate value
+    // QUIRK: in the 5th (and 10th) iteration of this loop it will overflow on the shift.
+    // This causes only the lower 4 bits to be shifted into place and removing the upper 3 bits
+    intermediate |= (byte & 0b01111111) << position;
+
+    if (position === 28) {
+      // Write to the view
+      U32_VIEW[0] = intermediate;
+      // set `intermediate` to the remaining 3 bits
+      // We only want the remaining three bits because the other 4 have been "consumed" on line 21
+      intermediate = (byte & 0b01110000) >>> 4;
+      // set `position` to -4 because later 7 will be added, making it 3
+      position = -4;
     }
-    if (!(byte & MSB)) return [decoded, i + 1];
+
+    position += 7;
+    i++;
+    // Keep going while there is a continuation bit
+  } while ((byte & 0b10000000) === 0b10000000);
+  const adjustedOffset = i - offset;
+  if (
+    (adjustedOffset === 10 && intermediate > -1) ||
+    adjustedOffset === 11 || i > buf.length
+  ) {
+    throw new RangeError("malformed or overflow varint");
   }
-  throw new RangeError("malformed or overflow varint");
+
+  // Write the intermediate value to the "empty" slot
+  // if the first slot is taken. Take the second slot
+  U32_VIEW[Number(adjustedOffset > 4)] = intermediate;
```

Comment:
```suggestion
  U32_VIEW[adjustedOffset > 4] = intermediate;
```

Reply:
I'll let Skye advise on this one, but I suspect this is another optimization. Should be noted in a comment if so.

Reply:
This `Number` looks just necessary to me (You can't use boolean for indicating first and second slot):

```
> let buf = new Uint32Array([0, 1]);
undefined
> buf[true]
undefined
> buf[Number(true)]
1
```

Reply:
Ah, yes, my misunderstanding. Nevermind.

Reply:
Ah completely misread this (thought it was a bitshift). Nevermind!